### PR TITLE
[Hardware] Support AMD (Rocm kernel)

### PR DESCRIPTION
**Pull request by yushengsu-thu**
This code base supports AMD GPUs
[Bugs Free]
- [Done] Fix (AMD) torch and Ray issue
- [Done] Add the conditions to enable (AMD) torch in the codebase
- [Done] AMD rocm Dockerfile
- [Done] Throughput metric - Tokens/Sec/GPU

[Pass test cases] `run_qwen2-7b_seq_balance.sh`, `run_qwen2-7b_rm_seq_balance.sh`
- [Done] Convergence test done 
- [Done] Throughput test done 

Tutorial:
- [Done] [AMD-doc](https://github.com/yushengsu-thu/verl/blob/main/docs/amd_tutorial)

Review Modification:
- [Done] Unified .gitignore
- [Done] Make support AMD parts to be the general case [AKA. `to.cuda()` --> `.to(torch.cuda.current_device())`]
- [Done] Ensure throughput and accuracy still remain the same on NV H100 GPUs with this version codebase.

Special thanks for the collaboration and help from:
(SGLang) @zhaochenyang20
(VeRL) @PeterSH6 
(AMD) @yushengsu-thu @vickytsang @xiaodoyu
(AnyScale) @hongpeng-guo @kevin85421

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,13 @@
+data
```
Comment (PeterSH6) [inline .gitignore]: These shall be deleted?

File: setup_amd.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    name='verl',
+    version=__version__,
+    package_dir={'': '.'},
+    packages=find_packages(where='.'),
+    url='https://github.com/volcengine/verl',
+    license='Apache 2.0',
+    author='Bytedance - Seed - MLSys',
+    author_email='zhangchi.usc1992@bytedance.com, gmsheng@connect.hku.hk',
+    description='veRL: Volcano Engine Reinforcement Learning for LLM',
+    # install_requires=install_requires,
```
Comment (PeterSH6) [inline setup_amd.py]: Just wondering if the setup file is ready? It seems that it didn't load the requirement_amd.txt

File: verl/third_party/vllm/__init__.py [hunk 3]
```diff
@@ -47,6 +47,16 @@ def get_version(pkg):
...
+elif package_version == '0.6.3+rocm624':
+    vllm_version = '0.6.3'
+    from .vllm_v_0_6_3_rocm624.llm import LLM
+    from .vllm_v_0_6_3_rocm624.llm import LLMEngine
+    from .vllm_v_0_6_3_rocm624 import parallel_state
+elif package_version == '0.6.4.dev4+gc9fc1608.rocm624':
+    vllm_version = '0.6.4'
+    from .vllm_v_0_6_4_rocm624.llm import LLM
+    from .vllm_v_0_6_4_rocm624.llm import LLMEngine
+    from .vllm_v_0_6_4_rocm624 import parallel_state
```
Comment (PeterSH6) [inline verl/third_party/vllm/__init__.py]: It seems that the relevant Rocm vllm is not uploaded yet.
I wonder if AMD supports vllm > 0.7 at the moment?

Comment (yushengsu-thu) [reply verl/third_party/vllm/__init__.py]: @PeterSH6. Guangming
1. It was ignored. Let me add it back.  (You can see it now)
2. At the same time, I'm testing vLLM > 0.7 because AMD seems to have not released a vLLM > 0.7 ready version. I am figuring out this and working with the AMD vLLM team now.

File: setup_amd.py (hunk 2 above)
Comment (yushengsu-thu) [reply setup_amd.py]: @PeterSH6 Guangming
If users want to run on AMD GPUs, they should follow this [AMD-doc](https://github.com/yushengsu-thu/verl/blob/main/docs/amd_tutorial/amd_example.txt) to install the VeRL. 
# Setup the Environment
```
pip install -r requirements_amd_no_deps.txt --no-deps
pip install -r requirements_amd.txt
```

I was thinking of merging this manual installation process into `setup.py` and make user can use command `pip install -e .` to finish installing:
(1) I can modify `setup.py` to have two parts: (1.1) detects AMD Torch and follows the installation process I provide, (1.2) and another that executes the original process in setup.py if AMD Torch is not detected.

(2) I previously tried to modify `setup.py` as described in (1), but it still executed the original setup.py. It seems likely that I also need to make some modifications to pyproject.toml to enable (1).

Do you think (1) is a good idea? Also, I’m not very familiar with solving (2). If you have time, could we quickly sync offline to figure out how to handle this part?

File: verl/third_party/vllm/__init__.py (hunk 3 above)
Comment (CZWin32768) [reply verl/third_party/vllm/__init__.py]: I compiled a vllm==0.7.3 on MI300, and tried it on verl. It works if I disable the sleep feature of vllm. But vllm has not supported sleep mode on amd devices.

[About AMD sleep mode support](https://github.com/vllm-project/vllm/pull/12695)

**Comment by CZWin32768** [comment]
When running Verl on ROCm devices, I encounter the following runtime error from `_get_device_properties`:
```
device >= 0 && device < num_gpus INTERNAL ASSERT FAILED at "../aten/src/ATen/hip/HIPContext.cpp":50, please report a bug to PyTorch. device=, num_gpus=.
``` 
This issue occurs on the following configurations: 
- lmsysorg/sglang:v0.4.3.post2-rocm630 (pytorch 2.6)
- rocm6.2 pytorch 2.6

There is a simple script to reproduce this error:
```
import os
import ray
import torch

ray.init(num_gpus=2)

@ray.remote(num_gpus=1)
class GPUActor:
    def ping(self):
        print('##### actor')
        print("GPU IDs: {}".format(ray.get_runtime_context().get_accelerator_ids()["GPU"]))
        print("ROCR_VISIBLE_DEVICES: {}".format(os.environ["ROCR_VISIBLE_DEVICES"]))
        if torch.cuda.is_available():
            print(f"Number of GPUs: {torch.cuda.device_count()}")
            for i in range(torch.cuda.device_count()):
                print(f"Device {i}: {torch.cuda.get_device_properties(i)}")
        else:
            print("No GPUs available")
        print('#####')

@ray.remote(num_gpus=1)
def gpu_task():
    print('##### task')
    print("GPU IDs: {}".format(ray.get_runtime_context().get_accelerator_ids()["GPU"]))
    print("ROCR_VISIBLE_DEVICES: {}".format(os.environ["ROCR_VISIBLE_DEVICES"]))
    if torch.cuda.is_available():
        print(f"Number of GPUs: {torch.cuda.device_count()}")
        for i in range(torch.cuda.device_count()):
            print(f"Device {i}: {torch.cuda.get_device_properties(i)}")
    else:
        print("No GPUs available")
    print('#####')

gpu_actor = GPUActor.remote()
ray.get(gpu_actor.ping.remote())
# The actor uses the first GPU so the task uses the second one.
ray.get(gpu_task.remote())
```

I found that if I only use tensors on each GPU without calling `get_device_properties`, the issue does not occur (this might be a workaround).  

I suspect this is why I can run verl2 with `vllm==0.7.3`, while `vllm==0.6.3` encounters the error since the v0.6.3 backend calls `get_device_properties`, triggering the assertion failure.

**Comment by zhaochenyang20** [comment]
@yushengsu-thu conflicts need to solve

**Comment by yushengsu-thu** [comment]
@PeterSH6 I've merged the conflict and updated the recommendation parts. Please kindly let me know if there are still issues. 

This code base supports AMD GPUs
[Bugs Free]

[Done] Fix (AMD) torch and Ray issue
[Done] Add the conditions to enable (AMD) torch in the codebase
[Done] AMD rocm Dockerfile
[Done] Throughput metric - Tokens/Sec/GPU
[Pass test cases] run_qwen2-7b_seq_balance.sh, run_qwen2-7b_rm_seq_balance.sh

[Done] Convergence test done
[Done] Throughput test done
Tutorial:

[Done] [AMD-doc](https://github.com/yushengsu-thu/verl/blob/main/docs/amd_tutorial)

File: verl/third_party/vllm/__init__.py (hunk 3 above)
Comment (yushengsu-thu) [reply verl/third_party/vllm/__init__.py]: @CZWin32768 , thank you for this try.

File: verl/third_party/vllm/__init__.py [hunk 4]
```diff
@@ -47,6 +47,11 @@ def get_version(pkg):
     from .vllm_v_0_6_3.llm import LLM
     from .vllm_v_0_6_3.llm import LLMEngine
     from .vllm_v_0_6_3 import parallel_state
+elif package_version == '0.6.3+rocm624':
+    vllm_version = '0.6.3'
+    from .vllm_v_0_6_3_rocm624.llm import LLM
+    from .vllm_v_0_6_3_rocm624.llm import LLMEngine
+    from .vllm_v_0_6_3_rocm624 import parallel_state
```
Comment (PeterSH6) [inline verl/third_party/vllm/__init__.py]: It seems that we can directly use `vllm_v_0_6_3` when using rocm624.

File: verl/third_party/vllm/vllm_v_0_6_3_rocm624/__init__.py
Comment (PeterSH6) [inline verl/third_party/vllm/vllm_v_0_6_3_rocm624/__init__.py:1]: So all the code relate to this third_party directory can be deleted

File: verl/trainer/ppo/ray_trainer.py [hunk 5]
```diff
@@ -334,6 +335,20 @@ def compute_timing_metrics(batch, timing_raw):
         },
     }
 
+###
+# [SUPPORT AMD:]
```
Comment (PeterSH6) [inline verl/trainer/ppo/ray_trainer.py]: It's a general feature. We can delete this comment

File: verl/trainer/ppo/ray_trainer.py [hunk 6]
```diff
@@ -993,6 +1009,14 @@ def fit(self):
                 # collect metrics
                 metrics.update(compute_data_metrics(batch=batch, use_critic=self.use_critic))
                 metrics.update(compute_timing_metrics(batch=batch, timing_raw=timing_raw))
+                
+                ###
+                # [SUPPORT AMD:]
+                config = self.config
```
Comment (PeterSH6) [inline verl/trainer/ppo/ray_trainer.py:1037]: Same here.

File: verl/workers/fsdp_workers.py [hunk 7]
```diff
@@ -405,15 +405,29 @@ def init_model(self):
...
             load_fsdp_model_to_gpu(self.actor_module_fsdp)
         if self._is_offload_optimizer:
             load_fsdp_optimizer(optimizer=self.actor_optimizer, device_id=torch.cuda.current_device())
 
-        data.batch = data.batch.cuda()
+        ###
+        # data.batch = data.batch.cuda()
+        # [SUPPORT AMD:]
+        if "AMD" in torch.cuda.get_device_name():
+            data.batch = data.batch.to(torch.cuda.current_device())
```
Comment (PeterSH6) [inline verl/workers/fsdp_workers.py]: It would be better if we can only keep `data.batch = data.batch.to(torch.cuda.current_device())` to support both AMD and NVIDIA (and other hardware)

Comment (PeterSH6) [reply verl/workers/fsdp_workers.py]: It seems that this line is mainly related the misalignment between AMD's torch and NVIDIA's torch

File: verl/third_party/vllm/__init__.py (hunk 4 above)
Comment (yushengsu-thu) [reply verl/third_party/vllm/__init__.py]: [Done]

File: verl/third_party/vllm/vllm_v_0_6_3_rocm624/__init__.py
Comment (yushengsu-thu) [reply verl/third_party/vllm/vllm_v_0_6_3_rocm624/__init__.py:1]: [Done]

File: verl/trainer/ppo/ray_trainer.py (hunk 5 above)
Comment (yushengsu-thu) [reply verl/trainer/ppo/ray_trainer.py]: [Done]

File: verl/trainer/ppo/ray_trainer.py (hunk 6 above)
Comment (yushengsu-thu) [reply verl/trainer/ppo/ray_trainer.py:1037]: [Done]

File: verl/workers/fsdp_workers.py (hunk 7 above)
Comment (yushengsu-thu) [reply verl/workers/fsdp_workers.py]: [Done]

File: requirements_amd_no_deps.txt [hunk 8]
```diff
@@ -0,0 +1,17 @@
+tensordict<0.6 #will need the coresponding torch version (need to ensure rocm support it)
```
Comment (PeterSH6) [inline requirements_amd_no_deps.txt]: Only keep this package in the dockerfile and delete this file

File: docker/Dockerfile.rocm [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    MAX_JOBS=$(nproc) python3 setup.py install && \
+    cd .. && \
+    rm -rf vllm
+
+# Copy the entire project directory
+COPY . .
+
+# Install dependencies
+RUN pip install -r requirements_amd_no_deps.txt --no-deps && \
+    pip install -r requirements_amd.txt && \
```
Comment (PeterSH6) [inline docker/Dockerfile.rocm]: Write the required package here in the dockerfile instead of using a requirements_amd.txt.
The requirements_amd.txt will be used in the setup.py

File: requirements_amd_no_deps.txt (hunk 8 above)
Comment (yushengsu-thu) [reply requirements_amd_no_deps.txt]: [Done]

File: docker/Dockerfile.rocm (hunk 9 above)
Comment (yushengsu-thu) [reply docker/Dockerfile.rocm]: [Done]

**Comment by PeterSH6** [comment]
The PR LGTM. 
The CI machine for AMD is not ready yet.
I think we can merge it in advance by making sure that these modification will not break other hardware? cc: @vermouth1992

File: setup_amd.py (hunk 2 above)
Comment (PeterSH6) [reply setup_amd.py]: Latest Status: The setup.py is not ready.
Users who want to try verl on AMD should use the Dockerfile provided by @yushengsu-thu

**Comment by vermouth1992** [comment]
LGTM.

File: setup_amd.py (hunk 2 above)
Comment (vermouth1992) [reply setup_amd.py]: Shall we create a readme on how to setup AMD GPU based environment in verl repo?

Comment (PeterSH6) [reply setup_amd.py]: Yes, there're two docs for AMD in docs/amd_tutorial/. Shall we add it in our main readme?
(Maybe a list of hardware support in the main readme)

File: verl/workers/rollout/vllm_rollout/__init__.py [hunk 10]
```diff
@@ -14,6 +14,11 @@
 
 from importlib.metadata import version, PackageNotFoundError
 
+###
+# [SUPPORT AMD:]
+import torch
```
Comment (eric-haibin-lin) [inline verl/workers/rollout/vllm_rollout/__init__.py:19]: @yushengsu-thu could you help explain why importing torch here is required?

File: verl/workers/fsdp_workers.py (hunk 7 above)
Comment (CZWin32768) [reply verl/workers/fsdp_workers.py]: > It would be better if we can only keep `data.batch = data.batch.to(torch.cuda.current_device())` to support both AMD and NVIDIA (and other hardware)

What if set cuda device with `torch.cuda.set_device` once, so that we can move data to cuda with `data = data.to('cuda')` for both AMD and NVIDIA devices?

File: verl/workers/rollout/vllm_rollout/__init__.py (hunk 10 above)
Comment (yushengsu-thu) [reply verl/workers/rollout/vllm_rollout/__init__.py:19]: @eric-haibin-lin , the following few-line codes:
```
if "AMD" in torch.cuda.get_device_name():
    import re
    ....
```
need to torch
